xmtp_id/src/associations/verified_signature.rs

```diff
@@ -100,6 +107,53 @@ impl VerifiedSignature {
         ))
     }
 
+    pub fn from_passkey(
+        public_key: &[u8],
+        signature: &[u8],
+        authenticator_data: &[u8],
+        client_data_json: &[u8],
+    ) -> Result<Self, SignatureError> {
+        // 1. Parse the public key from raw bytes
+        let verifying_key = match VerifyingKey::from_sec1_bytes(public_key) {
+            Ok(key) => key,
+            Err(_err) => return Err(SignatureError::InvalidPublicKey),
+        };
+
+        // 2. Parse the signature
+        let signature = match Signature::from_der(signature) {
+            Ok(sig) => sig,
+            Err(_) => return Err(SignatureError::Invalid),
+        };
+
+        // Get the origin from the client_data
+        let json: ClientDataJson = serde_json::from_slice(client_data_json)
+            .map_err(|_| SignatureError::MissingRelyingPartyOrigin)?;
+
+        // 3. Hash the client data
+        let mut hasher = Sha256::new();
+        hasher.update(client_data_json);
+        let client_data_hash = hasher.finalize();
+
+        // 4. Construct the verification data (authenticator_data + client_data_hash)
+        let mut verification_data =
+            Vec::with_capacity(authenticator_data.len() + client_data_hash.len());
+        verification_data.extend_from_slice(authenticator_data);
+        verification_data.extend_from_slice(&client_data_hash);
+
+        // 5. Verify the signature
+        verifying_key.verify(&verification_data, &signature)?;
```

Comment:
With all of our other signature verifications we use the verification text that was generated from the `SignatureRequest`.

It seems like the caller can pass in whatever `authenticator_data` and `client_data_json` it wants.

What is stopping someone from taking a valid passkey signature of some _totally unrelated_ payload (or another user's Identity Update) and passing it off as their own?

It seems like we need to verify that the `client_data_json` matches the expected verification text or re-generate our own `client_data_json` based on the `SignatureRequest` and have that passed in here.

Reply:
Got it.